File: algorithms/neetcode/algorithms/algorithms/src/main/java/com/mycompany/leetcode/blind75/LinkedList/RemoveNthNodeFromEndOfList/remove-nth-node-from-end-of-list-sgallivan.js
const { ListNode } = require('./ListNode');

const removeNthFromEnd = (head, n) => {
  let fast = head, slow = head;
  for(let i = 0; i < n; i++) fast = fast.next
  if(!fast) return head.next;
  while(fast.next) fast = fast.next, slow = slow.next
  slow.next = slow.next.next
  return head;
};

const head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));

console.log(head);

const result = removeNthFromEnd(head, 2);

console.log(result);

File: algorithms/neetcode/algorithms/algorithms/src/main/java/com/mycompany/leetcode/blind75/LinkedList/RemoveNthNodeFromEndOfList/remove-nth-node-from-end-of-list-sgallivan.py


class ListNode:
    def __init__(self, x, next = None):
        self.val = x
        self.next = next

class S:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        fast, slow = head, head
        for _ in range(n): fast = fast.next
        if not fast: return head.next
        while fast.next: fast, slow = fast.next, slow.next
        slow.next = slow.next.next
        return head 

head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))

result = S().removeNthFromEnd(head, 2)

print(result)

while result:
    print(result.val)
    tmp = result.next
    result = tmp

